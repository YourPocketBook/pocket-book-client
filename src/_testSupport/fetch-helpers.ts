import fetchMock, {
  MockOptionsMethodDelete,
  MockOptionsMethodGet,
  MockOptionsMethodPost,
  MockOptionsMethodPut,
  MockResponse,
  MockResponseFunction
} from "fetch-mock";

afterEach(fetchMock.reset);

export function authenticatedGet(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodGet
) {
  fetchMock.get(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}

export function authenticatedPut(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPut
) {
  fetchMock.put(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedPost(
  url: string,
  token: string,
  body: {},
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodPost
) {
  fetchMock.post(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}` &&
      h.body === JSON.stringify(body),
    response,
    options
  );
}

export function authenticatedDelete(
  url: string,
  token: string,
  response: MockResponse | MockResponseFunction,
  options?: MockOptionsMethodDelete
) {
  fetchMock.delete(
    (u, h) =>
      u === url &&
      (h.headers as { Authorization: string }).Authorization ===
        `Bearer ${token}`,
    response,
    options
  );
}
