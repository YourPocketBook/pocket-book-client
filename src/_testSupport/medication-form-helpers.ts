import { fireEvent } from "@testing-library/react";

export const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2019-12-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

export function setupState(getByLabelText: (text: string) => HTMLElement) {
  fireEvent.change(getByLabelText("Name"), {
    target: { value: testDefaultState.name }
  });
  fireEvent.change(getByLabelText("Policy Date"), {
    target: { value: testDefaultState.policyDate }
  });
  fireEvent.change(getByLabelText("Indications"), {
    target: { value: testDefaultState.indications }
  });
  fireEvent.change(getByLabelText("Inclusion Criteria"), {
    target: { value: testDefaultState.inclusionCriteria }
  });
  fireEvent.change(getByLabelText("Exclusion Criteria"), {
    target: { value: testDefaultState.exclusionCriteria }
  });
  fireEvent.change(getByLabelText("Name, Forms and Strengths of Medicine"), {
    target: { value: testDefaultState.form }
  });
  fireEvent.change(getByLabelText("Route or Method"), {
    target: { value: testDefaultState.route }
  });
  fireEvent.change(getByLabelText("Dosage"), {
    target: { value: testDefaultState.dose }
  });
  fireEvent.change(getByLabelText("Side Effects"), {
    target: { value: testDefaultState.sideEffects }
  });
  fireEvent.change(getByLabelText("Advice if Taken"), {
    target: { value: testDefaultState.adviceIfTaken }
  });
  fireEvent.change(getByLabelText("Advice if Declined"), {
    target: { value: testDefaultState.adviceIfDeclined }
  });
}

export function checkForClass(
  getByLabelText: (text: string) => HTMLElement,
  cls: "is-valid" | "is-invalid"
) {
  expect(getByLabelText("Name")).toHaveClass(cls);
  expect(getByLabelText("Policy Date")).toHaveClass(cls);
  expect(getByLabelText("Indications")).toHaveClass(cls);
  expect(getByLabelText("Inclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Exclusion Criteria")).toHaveClass(cls);
  expect(getByLabelText("Name, Forms and Strengths of Medicine")).toHaveClass(
    cls
  );
  expect(getByLabelText("Route or Method")).toHaveClass(cls);
  expect(getByLabelText("Dosage")).toHaveClass(cls);
  expect(getByLabelText("Side Effects")).toHaveClass(cls);
  expect(getByLabelText("Advice if Taken")).toHaveClass(cls);
  expect(getByLabelText("Advice if Declined")).toHaveClass(cls);
}

export function checkState(
  getByLabelText: (text: string) => HTMLElement,
  hasValue: boolean
) {
  expect(getByLabelText("Name")).toHaveValue(
    hasValue ? testDefaultState.name : ""
  );
  expect(getByLabelText("Policy Date")).toHaveValue(
    hasValue ? testDefaultState.policyDate : ""
  );
  expect(getByLabelText("Indications")).toHaveTextContent(
    hasValue ? testDefaultState.indications : ""
  );
  expect(getByLabelText("Inclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.inclusionCriteria : ""
  );
  expect(getByLabelText("Exclusion Criteria")).toHaveTextContent(
    hasValue ? testDefaultState.exclusionCriteria : ""
  );
  expect(
    getByLabelText("Name, Forms and Strengths of Medicine")
  ).toHaveTextContent(hasValue ? testDefaultState.form : "");
  expect(getByLabelText("Route or Method")).toHaveTextContent(
    hasValue ? testDefaultState.route : ""
  );
  expect(getByLabelText("Dosage")).toHaveTextContent(
    hasValue ? testDefaultState.dose : ""
  );
  expect(getByLabelText("Side Effects")).toHaveTextContent(
    hasValue ? testDefaultState.sideEffects : ""
  );
  expect(getByLabelText("Advice if Taken")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfTaken : ""
  );
  expect(getByLabelText("Advice if Declined")).toHaveTextContent(
    hasValue ? testDefaultState.adviceIfDeclined : ""
  );
}
