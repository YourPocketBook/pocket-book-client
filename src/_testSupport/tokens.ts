export const inDateNonAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.VmT5YkTG0sUd5sK7skfMdi9kc5XG7Yhx6sqhqMSxhVY";
export const inDateAdminToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE3MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInJvbGUiOiJBZG1pbiIsImdpdmVuX25hbWUiOiJUZXN0IFVzZXIiLCJuYW1laWQiOiJmZjFkMDhlNy00NjU2LTRjMjgtOTc2Zi0xYzI3ZjE5MTQ3N2YiLCJlbWFpbCI6InRlc3RAc2phLm9yZy51ayJ9.rSQb2Hv6RfxaXSfO6s3QjLjnYKeqVgps3rPenquh3Nk";
export const expiredToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNDE2MjM5MDIyLCJleHAiOjE0MTYyNDAwMjIsImF1ZCI6ImxvY2FsaG9zdCIsImlzcyI6ImxvY2FsaG9zdCIsInVuaXF1ZV9uYW1lIjoiVGVzdCBVc2VyIiwibmFtZWlkIjoiZmYxZDA4ZTctNDY1Ni00YzI4LTk3NmYtMWMyN2YxOTE0NzdmIiwiZW1haWwiOiJ0ZXN0QHNqYS5vcmcudWsifQ.7bDfHDk9sWXL8boFeZWIdo_GB-4Qj8PvA4HDC49EzNc";
