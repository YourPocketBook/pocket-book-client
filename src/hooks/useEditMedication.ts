import isValid from "date-fns/isValid";
import parseISO from "date-fns/parseISO";
import { useReducer } from "react";
import {
  IEditMedication,
  IEditMedicationValid
} from "../medications/medication-form";

export function useEditMedication() {
  const [state, dispatch] = useReducer(
    (
      s: IEditMedication & { valid: IEditMedicationValid },
      a: { prop: keyof IEditMedication } & (
        | { op: "update"; value: string | undefined | null }
        | {
            op: "invalidate";
          }
        | { op: "setValid"; value: boolean }
      )
    ) => {
      if (s.hasOwnProperty(a.prop)) {
        const newData = { ...s };

        switch (a.op) {
          case "update":
            newData[a.prop] = a.value || "";
            if (a.prop === "policyDate") {
              newData.valid[a.prop] = !!a.value && isValid(parseISO(a.value));
            } else {
              newData.valid[a.prop] = !!a.value && a.value.length > 0;
            }
            break;
          case "invalidate":
            newData.valid[a.prop] = false;
            break;
          case "setValid":
            newData.valid[a.prop] = a.value;
            break;
        }
        return newData;
      }
      return s;
    },
    {
      adviceIfDeclined: "",
      adviceIfTaken: "",
      dose: "",
      exclusionCriteria: "",
      form: "",
      inclusionCriteria: "",
      indications: "",
      name: "",
      policyDate: "",
      route: "",
      sideEffects: "",
      valid: {
        adviceIfDeclined: false,
        adviceIfTaken: false,
        dose: false,
        exclusionCriteria: false,
        form: false,
        inclusionCriteria: false,
        indications: false,
        name: false,
        policyDate: false,
        route: false,
        sideEffects: false
      }
    }
  );

  return { state, dispatch };
}
