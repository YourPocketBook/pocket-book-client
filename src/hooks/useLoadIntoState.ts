import { Dispatch, useEffect, useState } from "react";
import { fetcher } from "../fetcher";

/**
 * Fetches JSON from the specified uri and loads each property into individually editable state variables
 *
 * @param uri The URI to load the data from
 * @param setState An object containing setState functions for each property
 * @param [authenticate=false] If set to true, the current token will be passed with the call
 * @returns True if the request is currently running, otherwise false.
 */
export function useLoadIntoEditableState<
  TData extends { [key: string]: string | undefined | null } & object
>(
  uri: string,
  dispatch: Dispatch<{
    op: "update";
    prop: keyof TData;
    value: string | undefined | null;
  }>,
  authenticate = false
): boolean {
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchAbort = new AbortController();

    const options: RequestInit = {
      cache: "reload",
      method: "GET",
      signal: fetchAbort.signal
    };

    async function func() {
      try {
        const res = authenticate
          ? await fetcher.authFetch(uri, options)
          : await fetch(uri, options);

        if (res.ok) {
          const data = (await res.json()) as TData;
          for (const prop in data) {
            if (data.hasOwnProperty(prop)) {
              dispatch({ op: "update", prop, value: data[prop] });
            }
          }
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${uri}.`);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setLoading(false);
        }
      }
    }

    func();
    return () => fetchAbort.abort();
  }, [uri, authenticate, dispatch]);

  return loading;
}
