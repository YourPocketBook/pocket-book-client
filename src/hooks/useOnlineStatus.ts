// Included from https://github.com/21kb/react-hooks/blob/master/packages/react-online-status-hook
//
// Copyright Â© 2018 Kunall Banerjee
// MIT License

import { useEffect, useState } from "react";

export function useOnlineStatus() {
  const [state, setState] = useState(navigator.onLine);

  const onOnlineEvent = () => {
    setState(navigator.onLine);
  };

  const onOfflineEvent = () => {
    setState(navigator.onLine);
  };

  useEffect(() => {
    window.addEventListener("online", onOnlineEvent);
    window.addEventListener("offline", onOfflineEvent);

    return () => {
      window.removeEventListener("online", onOnlineEvent);
      window.removeEventListener("offline", onOfflineEvent);
    };
  });

  return state;
}
