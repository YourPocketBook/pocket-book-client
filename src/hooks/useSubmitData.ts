import { useState } from "react";
import { fetcher } from "../fetcher";

type IIsValid<T> = { [P in keyof T]: boolean };
type ISetIsValid<T> = { [P in keyof T]?: (val: boolean) => void };
type ISetCustomValid<T> = { [P in keyof T]?: (val: string) => void };

/**
 * Hook that handles submitting form data and responding to server-side validation.
 *
 * @param value Object that contains the current values to send to the server
 * @param valid Object that contains the validation status of each item in value
 * @param setValid Object that contains a set function for each item in valid
 * @param url The URI the data should be submitted to
 * @param options The options to control the submitting
 */
export function useSubmitData<TData extends object>(
  value: TData,
  valid: IIsValid<TData>,
  setValid: ISetIsValid<TData>,
  url: string,
  options: {
    setShowValidation?: any;
    method?: "PUT" | "POST";
    authenticate?: boolean;
    customVal?: ISetCustomValid<TData>;
    success?(arg?: {}): void;
  } = {}
) {
  const [serverError, setServerError] = useState(false);
  const [notAuthorised, setNotAuthorised] = useState(false);
  const [inProgress, setInProgress] = useState(false);

  async function submit(e: { preventDefault(): void }) {
    e.preventDefault();

    for (const prop in value) {
      if (value.hasOwnProperty(prop) && valid.hasOwnProperty(prop)) {
        if (!valid[prop]) {
          if (options.setShowValidation) {
            options.setShowValidation(true);
          }
          return;
        }
      }
    }

    if (options.setShowValidation) {
      options.setShowValidation(false);
    }

    setServerError(false);
    setNotAuthorised(false);
    setInProgress(true);

    const method = options.method ? options.method : "POST";

    const requestOptions = {
      body: JSON.stringify(value),
      headers: {
        "Content-Type": "application/json"
      },
      method
    };

    try {
      const res = options.authenticate
        ? await fetcher.authFetch(url, requestOptions)
        : await fetch(url, requestOptions);

      if (res.ok) {
        if (options.success) {
          if (res.status !== 204) {
            options.success(await res.json());
          } else {
            options.success();
          }
        }
      } else {
        if (options.setShowValidation) {
          options.setShowValidation(true);
        }

        if (res.status === 400 || res.status === 409 || res.status === 403) {
          const body = (await res.json()) as Array<{
            path: keyof TData;
            error: "isBlank" | "isInvalid";
          }>;

          for (const b of body) {
            if (b.error === "isBlank" || b.error === "isInvalid") {
              const validFunc = setValid[b.path];

              if (validFunc) {
                validFunc(false);
              }
            } else if (options.customVal) {
              const func = options.customVal[b.path];
              if (func) {
                func(b.error);
              }
            }
          }
        } else if (res.status === 401) {
          setNotAuthorised(true);
        } else {
          setServerError(true);
        }
      }
    } finally {
      setInProgress(false);
    }
  }

  return { submit, serverError, notAuthorised, inProgress };
}
