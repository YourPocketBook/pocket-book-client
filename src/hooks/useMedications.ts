import { useEffect, useState } from "react";
import { fetcher } from "../fetcher";
import { IMedication } from "../model/medication";
import { useOnlineStatus } from "./useOnlineStatus";

const medicationsKey = "medications";
const listEndpoint = "/api/get-medications";
const detailEndpoint = "/api/get-medication";

export function useMedicationList() {
  const [isLoading, setIsLoading] = useState(false);
  const [serverError, setServerError] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const localData = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : [];
  const [medicationList, setMedicationList] = useState(localData);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);

    async function func() {
      try {
        const res = await fetcher.authFetch(listEndpoint, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const data = (await res.json()) as Array<Partial<IMedication>>;

          if (fetchAbort.signal.aborted) {
            return;
          }

          const cacheString = localStorage.getItem(medicationsKey);
          const oldData = cacheString
            ? (JSON.parse(cacheString) as Array<Partial<IMedication>>)
            : [];

          const currentIds = data
            .filter(m => m && m.id)
            .map(m => m.id as number);
          const previousIds = oldData
            .filter(m => m && m.id)
            .map(m => m.id as number);

          const updatedItems = oldData
            .filter(m => currentIds.includes(m.id as number))
            .map(m => {
              const update = data.find(n => n.id === m.id);
              if (update && update.name) {
                m.name = update.name;
              }
              return m;
            })
            .concat(data.filter(m => !previousIds.includes(m.id as number)));

          setMedicationList(updatedItems);
          localStorage.setItem(medicationsKey, JSON.stringify(updatedItems));
          setUpdateAvailable(true);
          setServerError(false);
        } else {
          setServerError(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          setServerError(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline]);

  return {
    isLoading,
    medications: medicationList
      .filter(m => m && m.id)
      .map(m => ({
        availableOffline: !!m.inclusionCriteria,
        id: m.id as number,
        name: m.name as string
      })),
    serverError,
    updateAvailable
  };
}

export function useMedication(id: string, refresh: boolean) {
  const [isLoading, setIsLoading] = useState(false);
  const [updateAvailable, setUpdateAvailable] = useState(false);
  const [loadFailed, setLoadFailed] = useState(false);
  const isOnline = useOnlineStatus();

  const localStoreString = localStorage.getItem(medicationsKey);
  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;
  const med = data ? data.find(m => m.id === parseInt(id, 10)) : undefined;
  const [medication, setMedication] = useState(med);

  useEffect(() => {
    if (!isOnline) {
      return;
    }

    const fetchAbort = new AbortController();

    const options = {
      cache: refresh ? ("no-cache" as "no-cache") : undefined,
      method: "GET",
      signal: fetchAbort.signal
    };

    setIsLoading(true);
    setLoadFailed(false);

    async function func() {
      try {
        const res = await fetcher.authFetch(`${detailEndpoint}/${id}`, options);

        if (fetchAbort.signal.aborted) {
          return;
        }

        if (res.ok) {
          const newData = (await res.json()) as IMedication;

          if (fetchAbort.signal.aborted) {
            return;
          }

          setMedication(newData);
          cacheMedication(newData);
          setUpdateAvailable(true);
        } else {
          setLoadFailed(true);
        }
      } catch (ex) {
        if (!fetchAbort.signal.aborted) {
          // tslint:disable-next-line:no-console
          console.log(`Could not get from ${detailEndpoint}/${id}`);
          setLoadFailed(true);
        }
      } finally {
        if (!fetchAbort.signal.aborted) {
          setIsLoading(false);
        }
      }
    }

    func();

    return () => fetchAbort.abort();
  }, [isOnline, id, refresh]);

  return {
    isLoading,
    loadFailed,
    medication,
    updateAvailable
  };
}

export function cacheMedication(medication: IMedication) {
  const localStoreString = localStorage.getItem(medicationsKey);

  const data = localStoreString
    ? (JSON.parse(localStoreString) as Array<Partial<IMedication>>)
    : undefined;

  let newList: Array<Partial<IMedication>>;

  if (data) {
    newList = data.map(m => (m.id === medication.id ? medication : m));
  } else {
    newList = [medication];
  }

  localStorage.setItem(medicationsKey, JSON.stringify(newList));
}
