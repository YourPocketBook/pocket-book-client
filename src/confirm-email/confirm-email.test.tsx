import "@testing-library/jest-dom/extend-expect";
import { render, wait } from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { ConfirmEmail } from ".";
import "../_testSupport/fetch-helpers";
import { useOnlineStatus } from "../hooks/useOnlineStatus";

jest.mock("../hooks/useOnlineStatus");

const testUserId = "TestUser";
const testCode = "123456";
const endpoint = "/api/confirm-user";

describe("Confirm Email", () => {
  beforeEach(() => {
    (useOnlineStatus as jest.Mock).mockReturnValue(true);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<ConfirmEmail userId={testUserId} code={testCode} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("displays initial state correctly", async () => {
    fetchMock.post(
      `${endpoint}?userId=${testUserId}&code=${testCode}`,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { container } = render(
      <ConfirmEmail userId={testUserId} code={testCode} />
    );

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("displays confirmation after calling the API and succeeding", async () => {
    fetchMock.post(`${endpoint}?userId=${testUserId}&code=${testCode}`, {
      status: 204
    });

    const { container } = render(
      <ConfirmEmail userId={testUserId} code={testCode} />
    );

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("displays error after calling the API and failing", async () => {
    fetchMock.post(`${endpoint}?userId=${testUserId}&code=${testCode}`, {
      body: {
        error: "isInvalid",
        path: "code"
      },
      status: 400
    });

    const { container } = render(
      <ConfirmEmail userId={testUserId} code={testCode} />
    );

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("displays error after calling the API when offline", async () => {
    (useOnlineStatus as jest.Mock).mockReturnValue(false);

    const { container } = render(
      <ConfirmEmail userId={testUserId} code={testCode} />
    );

    expect(container).toMatchSnapshot();
  });
});
