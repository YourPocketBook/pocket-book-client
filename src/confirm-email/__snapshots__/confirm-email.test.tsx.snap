// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm Email displays confirmation after calling the API and succeeding 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <h1
              class="header1"
            >
              Email Address Confirmed
            </h1>
            <p
              class="main-text"
            >
              Thanks, you have confirmed your email address. You can now sign in.
            </p>
            <a
              class="btn btn-dark-green"
              href="/"
              id="home-button"
            >
              Home
            </a>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Confirm Email displays error after calling the API and failing 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <h1
              class="header1"
            >
              Failed to Confirm
            </h1>
            <p
              class="main-text"
            >
              Sorry, that hasn't worked for some reason. Try the link in your email again, and if it fails again, it might be too old. Try to log in, you will be sent a fresh link.
            </p>
            <a
              class="btn btn-dark-green"
              href="/"
              id="home-button"
            >
              Home
            </a>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Confirm Email displays error after calling the API when offline 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <div
              class="alert alert-dark fade"
              role="alert"
            >
              You appear to be offline.
            </div>
            <h1
              class="header1"
            >
              Cannot Confirm - Offline
            </h1>
            <p
              class="main-text"
            >
              You seem to be offline.  Go online and we will try again.
            </p>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Confirm Email displays initial state correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
    </nav>
    <div
      class="centreContainer container"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-6 offset-lg-3"
        >
          <div
            class="centreBox"
          >
            <h1
              class="header1"
            >
              Confirming Email Address
            </h1>
            <p
              class="main-text"
            >
              Thanks for confirming your email, we are just updating our records.  Give us a moment please.
            </p>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
